source/common/common/shared_hash_map.h

```diff
@@ -0,0 +1,373 @@
+#pragma once
+
+#include <pthread.h>
+
+#include <algorithm>
+#include <string>
+
+#include "common/common/assert.h"
+#include "common/common/hash.h"
+#include "common/common/logger.h"
+
+#include "absl/base/thread_annotations.h"
+#include "absl/strings/string_view.h"
+#include "fmt/format.h"
+
+namespace Envoy {
+
+/**
+ * Initialization parameters for SharedHashMap. The options are duplicated
+ * to the control-block after init, to aid with sanity checking when attaching
+ * an existing memory segment.
+ */
+struct SharedHashMapOptions {
+  std::string toString() const {
+    return fmt::format("capacity={}, max_key_size={}, num_slots={}", capacity, max_key_size,
+                       num_slots);
+  }
+  bool operator==(const SharedHashMapOptions& that) const {
+    return capacity == that.capacity && max_key_size == that.max_key_size &&
+           num_slots == that.num_slots;
+  }
+  bool operator!=(const SharedHashMapOptions& that) const { return !(*this == that); }
+
+  uint32_t capacity;     // how many values can be stored.
+  uint32_t max_key_size; // how many bytes of string can be stored.
+  uint32_t num_slots;    // determines speed of hash vs size efficiency.
+};
+
+/**
+ * Implements hash_map<string, Value> without using pointers, suitable
+ * for use in shared memory. Users must commit to capacity, max_key_size,
+ * and num_slots at construction time, and to the payload value at compile-time.
+ *
+ * This map may also be suitable for persisting a hash-table to long term storage,
+ * but not across machine architectures, as it doesn't use network byte order for
+ * storing ints.
+ */
+template <class Value> class SharedHashMap : public Logger::Loggable<Logger::Id::config> {
+public:
+  /**
+   * Sentinal used for next_cell links to indicate end-of-list.
+   */
+  static const uint32_t Sentinal = 0xffffffff;
+
+  /**
+   * Constructs a map control structure given a set of options, which cannot be changed.
+   * Note that the map dtaa itself is not constructed when the object is constructed.
+   * After the control-structure is constructed, the number of bytes can be computed so
+   * that a shared-memory segment can be allocated and passed to init() or attach().
+   */
+  SharedHashMap(const SharedHashMapOptions& options)
+      : options_(options), cells_(nullptr), control_(nullptr), slots_(nullptr) {}
+
+  /**
+   * Returns the numbers of byte required for the hash-table, based on
+   * the control structure. This must be used to allocate the
+   * backing-store (eg) in shared memory, which we do after
+   * constructing the object with the desired sizing.
+   */
+  size_t numBytes() const {
```

Comment:
I think the interface might be cleaner if you make this a static function that takes the Options, and then put the code in attach() into the constructor.

Reply:
I'm still figuring out how to integrate this.  Making this a static function is fine, but it looks like the caller will know ahead of time whether they are attempting to attach or just init, so I'm reluctant to attempt attaching in the ctor.

Reply:
I'm not 100% sure this made it cleaner, but it's reasonable.